### ISSUE TITLE
tf.function decorator with GradientTape >10x slower than tf.keras.model.Models.fit()

### ISSUE THREAD

**Issue by gaborchris** [tensorflow/tensorflow#31938]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No, reused code from https://www.tensorflow.org/beta/guide/effective_tf2 with minor modifications 
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab (originally found on Ubuntu 18.04)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): 
- TensorFlow version (use command below): 2
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When using @tf.function decorator with  tf.GradientTape to compute updates on a model, I find that it takes 110 seconds per epoch. When I used model.fit(dataset), it only takes ~7 seconds. I'm not sure why it takes so much longer using the custom train function. This is code copied over from the Effective TensorFlow 2 documentation on tensorflow.org. To make sure this not specific to my personal machine, I used google colab with GPU enabled under Notebook settings. 

**Describe the expected behavior**
I would expect that using @tf.function decorator for the train function would compute the gradient update steps in a similar amount of time as model.fit(). I am wondering if the model is not being put onto GPU and is kept on CPU when trained this way.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

try:
  # %tensorflow_version only exists in Colab.
  %tensorflow_version 2.x
except Exception:
  pass

import tensorflow as tf
import time

class MyModel(tf.keras.models.Model):

    def __init__(self):
        super(MyModel, self).__init__()
        self.conv1 = tf.keras.layers.Conv2D(32, 3, activation='relu')
        self.flatten = tf.keras.layers.Flatten()
        self.d1 = tf.keras.layers.Dense(128, activation='relu')
        self.d2 = tf.keras.layers.Dense(10, activation='softmax')
        self.times_called = tf.Variable(0.0, trainable=False)

    def call(self, x):
        self.times_called.assign_add(1)
        x = self.conv1(x)
        x = self.flatten(x)
        x = self.d1(x)
        return self.d2(x)

optimizer = tf.keras.optimizers.Adam()
loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
@tf.function
def train_step(model, dataset):
    for images, labels in dataset:
        with tf.GradientTape() as tape:
            preds = model(images)
            loss = loss_object(labels, preds)
        gradients = tape.gradient(loss, model.trainable_variables)
        optimizer.apply_gradients(zip(gradients, model.trainable_variables))

if __name__ == "__main__":

    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0
    train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(32)

    model = MyModel()
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy')
    model.fit(train_ds)

    model2 = MyModel()
    start = time.time()
    train_step(model2, train_ds)
    print(model.times_called)
    print(time.time() - start)

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

This is the output I get when running this code on google colab:
1563/1563 [==============================] - 7s 4ms/step - loss: 1.4598
<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=1563.0>
110.49453139305115

**Comment by gaborchris** [comment]
`

**Comment by DEKHTIARJonathan** [comment]
@tf.function should not be around the whole epoch: `for images, labels in dataset:`
Move the loop outside the function and it will fix your issue

**Comment by akanyaani** [comment]
Hi gaborchris,
As suggested by DEKHTIARJonathan change your code like this it will be faster than model.fit.


```
@tf.function
def train_step(model, image, labels):
    with tf.GradientTape() as tape:
        preds = model(images)
        loss = loss_object(labels, preds)
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```

```
for images, labels in train_ds:
        train_step(model2, images, labels)
```
```
1563/1563 [==============================] - 7s 4ms/step - loss: 1.4690
<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=1563.0>
5.470849275588989
```

**Comment by oanush** [comment]
@gaborchris ,
Can you please try the solution provided by @DEKHTIARJonathan and @akanyaani.Thanks!
